File: scripts/audit.py
#!/usr/bin/env python3
# audit.py — Comprehensive pre-refactor readiness audit
# Combines health, coverage, drift, mutation risk, and chain status
# into a single report.
#
# Usage:
#   python scripts/audit.py
#   python scripts/audit.py --strict   (exit 1 if any issues found)
#
# This addresses the gap where agents had to manually run health + coverage + drift
# before deciding it was safe to refactor. Audit combines all checks and gives
# a clear YES/NO answer.

import json
import os
import sys
import subprocess


def run_command(cmd):
    """Run a command and return its output."""
    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True, timeout=120,
            cwd=os.getcwd()
        )
        return result.stdout, result.returncode
    except Exception as e:
        return str(e), 1


def check_manifest():
    """Check if manifest exists and is valid."""
    manifest_path = os.path.join(os.getcwd(), 'regrets', 'manifest.json')
    if not os.path.exists(manifest_path):
        return False, "No manifest.json found"
    try:
        with open(manifest_path, 'r') as f:
            manifest = json.load(f)
        clusters = manifest.get('clusters', [])
        if not clusters:
            return False, "No clusters defined in manifest"
        return True, f"{len(clusters)} cluster(s) defined"
    except json.JSONDecodeError as e:
        return False, f"Invalid JSON: {e}"


def check_regret_files():
    """Check if .regret files exist for all clusters."""
    regret_dir = os.path.join(os.getcwd(), 'regrets')
    manifest_path = os.path.join(regret_dir, 'manifest.json')

    with open(manifest_path, 'r') as f:
        manifest = json.load(f)

    cluster_ids = {c['id'] for c in manifest.get('clusters', [])}
    regret_files = set()
    try:
        for f in os.listdir(regret_dir):
            if f.endswith('.regret'):
                regret_files.add(os.path.splitext(f)[0])
    except FileNotFoundError:
        return False, "No regrets/ directory"

    missing = cluster_ids - regret_files
    if missing:
        return False, f"Missing .regret files: {', '.join(sorted(missing))}"
    return True, f"All {len(cluster_ids)} .regret files present"


def check_validate():
    """Run validate and check all GREEN."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output, code = run_command(['python3', f'{script_dir}/validate.py'])
    if code == 0:
        # Count GREEN clusters
        green_count = output.count('✅')
        return True, f"{green_count} cluster(s) GREEN"
    else:
        red_count = output.count('❌')
        return False, f"{red_count} cluster(s) RED — fix before refactoring"


def check_drift():
    """Run drift detection (3 runs)."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output, code = run_command(['python3', f'{script_dir}/validate.py', '--runs', '3'])
    if 'DRIFT' in output:
        drift_count = output.count('DRIFT')
        return False, f"{drift_count} cluster(s) DRIFT — add normalize rules"
    if code == 0:
        stable_count = output.count('STABLE')
        return True, f"{stable_count} cluster(s) STABLE across 3 runs"
    return False, "Drift check failed"


def check_health():
    """Parse health report for SOLID clusters."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output, code = run_command(['python3', f'{script_dir}/health.py'])

    solid_count = output.count('SOLID')
    good_count = output.count('GOOD')
    unstable_count = output.count('UNSTABLE')
    fragile_count = output.count('FRAGILE')

    if fragile_count > 0 or unstable_count > 0:
        return False, f"SOLID: {solid_count}, GOOD: {good_count}, UNSTABLE: {unstable_count}, FRAGILE: {fragile_count}"
    return True, f"SOLID: {solid_count}, GOOD: {good_count}"


def check_coverage():
    """Run coverage check (Python clusters only)."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output, code = run_command(['python3', f'{script_dir}/coverage.py'])

    if 'LOW' in output:
        low_count = output.count('LOW')
        return False, f"{low_count} cluster(s) with LOW coverage"
    if 'PARTIAL' in output:
        partial_count = output.count('PARTIAL')
        return True, f"All clusters covered, {partial_count} PARTIAL (acceptable but improve)", True  # has_warning
    if code == 0:
        return True, "All clusters FULL or LIKELY FULL"
    return True, "Coverage check skipped (non-Python clusters)"


def check_mutation_risk():
    """Check for clusters with single input — mutation risk unverifiable.

    Clusters with only 1 input cannot verify mutation behavior because
    there is no second input to compare against. If a function mutates
    its input, the single test won't catch it. This is a manifest-level
    heuristic check (no static analysis required).

    Additionally checks for duplicate inputs in the inputs array —
    if inputs[0] and inputs[1] are identical, they're not providing
    independent verification.
    """
    manifest_path = os.path.join(os.getcwd(), 'regrets', 'manifest.json')
    if not os.path.exists(manifest_path):
        return True, "No manifest — mutation risk check skipped"

    try:
        with open(manifest_path, 'r') as f:
            manifest = json.load(f)
    except json.JSONDecodeError:
        return True, "Invalid manifest — mutation risk check skipped"

    clusters = manifest.get('clusters', [])
    if not clusters:
        return True, "No clusters — mutation risk check skipped"

    single_input = []
    duplicate_inputs = []

    for c in clusters:
        inputs = c.get('inputs', [])
        input_count = len(inputs)

        if input_count <= 1:
            single_input.append(c['id'])
        elif input_count >= 2:
            # Check for duplicate inputs
            try:
                serialized = [json.dumps(inp, sort_keys=True, ensure_ascii=False) for inp in inputs]
                if len(serialized) != len(set(serialized)):
                    duplicate_inputs.append(c['id'])
            except (TypeError, ValueError):
                pass  # Can't serialize — skip duplicate check

    parts = []
    if single_input:
        parts.append(f"{len(single_input)} cluster(s) have single input — mutation risk unverifiable")
    if duplicate_inputs:
        parts.append(f"{len(duplicate_inputs)} cluster(s) have duplicate inputs — not independently verified")

    if parts:
        msg = '; '.join(parts)
        # This is a warning, not a hard fail — single input is valid but risky
        return True, msg, True  # has_warning

    return True, "All clusters have ≥2 unique inputs — mutation risk verifiable"


def check_truth():
    """Check if dual truth baselines (KEBENARAN 1 and KEBENARAN 2) have been captured."""
    proof_dir = os.path.join(os.getcwd(), 'proof')
    
    # Find project-specific subdirectory
    manifest_path = os.path.join(os.getcwd(), 'regrets', 'manifest.json')
    project_name = os.getcwd().split('/')[-1]  # default to cwd name
    if os.path.exists(manifest_path):
        try:
            with open(manifest_path, 'r') as f:
                manifest = json.load(f)
            project_name = manifest.get('projectName', project_name)
        except (json.JSONDecodeError, IOError):
            pass
    
    project_proof_dir = os.path.join(proof_dir, project_name)
    k1_path = os.path.join(project_proof_dir, 'KEBENARAN_1_raw_output.json')
    k2_path = os.path.join(project_proof_dir, 'KEBENARAN_2_fingerprints.json')
    
    if not os.path.exists(k1_path) or not os.path.exists(k2_path):
        return False, "KEBENARAN baselines not captured — run 'regret truth' first"
    
    try:
        with open(k1_path, 'r') as f:
            k1 = json.load(f)
        with open(k2_path, 'r') as f:
            k2 = json.load(f)
        
        k1_count = len(k1)
        k2_count = len(k2.get('fingerprints', {}))
        k2_chains = len(k2.get('chains', {}))
        
        if k1_count != k2_count:
            return False, f"K1 has {k1_count} clusters, K2 has {k2_count} fingerprints — MISMATCH"
        
        chain_info = f", {k2_chains} chains" if k2_chains > 0 else ""
        return True, f"Both truths captured: {k1_count} clusters{chain_info}"
    except (json.JSONDecodeError, IOError) as e:
        return False, f"Error reading truth files: {e}"


def main():
    strict = '--strict' in sys.argv

    print("╔══════════════════════════════════════════════════════════════════╗")
    print("║              REGRET PRE-REFACTOR AUDIT REPORT                   ║")
    print("╚══════════════════════════════════════════════════════════════════╝")
    print()

    checks = [
        ("Manifest", check_manifest),
        ("Regret Files", check_regret_files),
        ("Validation", check_validate),
        ("Drift Detection", check_drift),
        ("Cluster Health", check_health),
        ("Branch Coverage", check_coverage),
        ("Mutation Risk", check_mutation_risk),
        ("Dual Truth", check_truth),
    ]

    all_pass = True
    warnings = []

    for name, check_fn in checks:
        print(f"  Checking {name}...", end=" ", flush=True)
        try:
            result = check_fn()
            if len(result) == 3:
                ok, msg, has_warning = result
            else:
                ok, msg = result
                has_warning = False

            if ok:
                icon = "✅" if not has_warning else "🟡"
                print(f"{icon} {msg}")
                if has_warning:
                    warnings.append(f"{name}: {msg}")
            else:
                print(f"❌ {msg}")
                all_pass = False
        except Exception as e:
            print(f"⚠️  Error: {e}")
            warnings.append(f"{name}: check failed — {e}")

    print()
    print("─" * 68)

    if all_pass and not warnings:
        print("✅ AUDIT PASSED — All checks GREEN. Safe to refactor.")
    elif all_pass and warnings:
        print("🟡 AUDIT PASSED with warnings:")
        for w in warnings:
            print(f"   • {w}")
        print("   Refactoring is possible but consider addressing warnings first.")
    else:
        print("❌ AUDIT FAILED — Fix the issues above before refactoring.")

    print()

    if strict and not all_pass:
        sys.exit(1)


if __name__ == '__main__':
    main()

File: scripts/audit.js
#!/usr/bin/env node
// audit.js — Comprehensive pre-refactor readiness audit (JS/TS stack)
// Combines health, coverage, drift, and chain status into a single report.
// This is the JS/TS equivalent of audit.py — it addresses the gap where
// JS/TS projects had no combined pre-refactor readiness check.
//
// Usage:
//   node scripts/audit.js
//   node scripts/audit.js --strict   (exit 1 if any issues found)
//
// Inspired by the Coretax-Auto-Downloader project — a Chrome extension
// where running health + coverage + drift manually before refactoring
// was error-prone and time-consuming.

import { readFileSync, existsSync, readdirSync } from 'fs'
import { resolve, join } from 'path'
import { execFileSync } from 'child_process'

const args = process.argv.slice(2)
const strict = args.includes('--strict')
const CWD = process.cwd()
const REGRET_DIR = resolve(CWD, 'regrets')
const MANIFEST_PATH = resolve(CWD, 'regrets/manifest.json')

// ─── Helper ────────────────────────────────────────────────────────────────────

function runCommand(cmd, cmdArgs) {
  try {
    const result = execFileSync(cmd, cmdArgs, {
      encoding: 'utf8',
      cwd: CWD,
      timeout: 120000,
      stdio: ['pipe', 'pipe', 'pipe']
    })
    return { output: result, code: 0 }
  } catch (err) {
    return { output: err.stdout || '', code: err.status || 1 }
  }
}

// ─── Checks ────────────────────────────────────────────────────────────────────

function checkManifest() {
  if (!existsSync(MANIFEST_PATH)) {
    return { ok: false, msg: 'No manifest.json found' }
  }
  try {
    const manifest = JSON.parse(readFileSync(MANIFEST_PATH, 'utf8'))
    const clusters = manifest.clusters || []
    if (!clusters.length) {
      return { ok: false, msg: 'No clusters defined in manifest' }
    }
    // Check for common manifest issues
    const issues = []
    for (const cluster of clusters) {
      if (!cluster.id) issues.push(`cluster missing "id" field`)
      if (!cluster.entry) issues.push(`cluster "${cluster.id}" missing "entry" field`)
      if (!cluster.file) issues.push(`cluster "${cluster.id}" missing "file" field`)
    }
    if (issues.length > 0) {
      return { ok: false, msg: `${issues.length} issue(s): ${issues.slice(0, 3).join('; ')}` }
    }
    return { ok: true, msg: `${clusters.length} cluster(s) defined` }
  } catch (e) {
    return { ok: false, msg: `Invalid JSON: ${e.message}` }
  }
}

function checkRegretFiles() {
  if (!existsSync(REGRET_DIR)) {
    return { ok: false, msg: 'No regrets/ directory' }
  }
  try {
    const manifest = JSON.parse(readFileSync(MANIFEST_PATH, 'utf8'))
    const clusterIds = new Set((manifest.clusters || []).map(c => c.id))
    const regretFiles = new Set()
    for (const f of readdirSync(REGRET_DIR)) {
      if (f.endsWith('.regret')) regretFiles.add(f.replace('.regret', ''))
    }
    const missing = [...clusterIds].filter(id => !regretFiles.has(id))
    if (missing.length > 0) {
      return { ok: false, msg: `Missing .regret files: ${missing.join(', ')}` }
    }
    return { ok: true, msg: `All ${clusterIds.size} .regret files present` }
  } catch (e) {
    return { ok: false, msg: `Error: ${e.message}` }
  }
}

function checkValidate() {
  const scriptDir = resolve(import.meta.dirname || '.', '')
  const validatePath = resolve(scriptDir, 'validate.js')
  if (!existsSync(validatePath)) {
    return { ok: false, msg: 'validate.js not found' }
  }
  const { output, code } = runCommand('node', [validatePath])
  if (code === 0) {
    const greenCount = (output.match(/✅/g) || []).length
    return { ok: true, msg: `${greenCount} cluster(s) GREEN` }
  } else {
    const redCount = (output.match(/❌/g) || []).length
    return { ok: false, msg: `${redCount} cluster(s) RED — fix before refactoring` }
  }
}

function checkDrift() {
  const scriptDir = resolve(import.meta.dirname || '.', '')
  const validatePath = resolve(scriptDir, 'validate.js')
  if (!existsSync(validatePath)) {
    return { ok: false, msg: 'validate.js not found for drift check' }
  }
  const { output, code } = runCommand('node', [validatePath, '--runs', '3'])
  if (output.includes('DRIFT')) {
    const driftCount = (output.match(/DRIFT/g) || []).length
    return { ok: false, msg: `${driftCount} cluster(s) DRIFT — add normalize rules` }
  }
  if (code === 0) {
    const stableCount = (output.match(/STABLE/g) || []).length
    return { ok: true, msg: `${stableCount} cluster(s) STABLE across 3 runs` }
  }
  return { ok: false, msg: 'Drift check failed' }
}

function checkHealth() {
  const scriptDir = resolve(import.meta.dirname || '.', '')
  const healthPath = resolve(scriptDir, 'health.js')
  if (!existsSync(healthPath)) {
    return { ok: false, msg: 'health.js not found' }
  }
  const { output } = runCommand('node', [healthPath])
  const solidCount = (output.match(/SOLID/g) || []).length
  const goodCount = (output.match(/GOOD/g) || []).length
  const unstableCount = (output.match(/UNSTABLE/g) || []).length
  const fragileCount = (output.match(/FRAGILE/g) || []).length

  if (fragileCount > 0 || unstableCount > 0) {
    return { ok: false, msg: `SOLID: ${solidCount}, GOOD: ${goodCount}, UNSTABLE: ${unstableCount}, FRAGILE: ${fragileCount}` }
  }
  return { ok: true, msg: `SOLID: ${solidCount}, GOOD: ${goodCount}` }
}

function checkCoverage() {
  const scriptDir = resolve(import.meta.dirname || '.', '')
  const coveragePath = resolve(scriptDir, 'coverage.js')
  if (!existsSync(coveragePath)) {
    return { ok: true, msg: 'Coverage check skipped (coverage.js not found)', warning: true }
  }
  const { output, code } = runCommand('node', [coveragePath])
  if (output.includes('UNDER-COVERED')) {
    const underCount = (output.match(/UNDER-COVERED/g) || []).length
    return { ok: false, msg: `${underCount} cluster(s) UNDER-COVERED` }
  }
  if (output.includes('PARTIAL')) {
    const partialCount = (output.match(/PARTIAL/g) || []).length
    return { ok: true, msg: `All clusters covered, ${partialCount} PARTIAL (acceptable but improve)`, warning: true }
  }
  return { ok: true, msg: 'All clusters WELL-COVERED' }
}

function checkMutationRisks() {
  // Detect JS/TS functions that mutate their input arguments.
  // This fills the gap where Python has mutate_audit.py but JS had nothing.
  // Common patterns: delete obj.prop, obj.prop = value on function arguments,
  // Object.assign(target, ...), array .push/.splice/.sort on args.
  const manifest = JSON.parse(readFileSync(MANIFEST_PATH, 'utf8'))
  const clusters = manifest.clusters || []
  const risks = []

  for (const cluster of clusters) {
    if (cluster.stack === 'python') continue  // Python has mutate_audit.py
    const filePath = resolve(CWD, cluster.file)
    if (!existsSync(filePath)) continue

    let source
    try {
      source = readFileSync(filePath, 'utf8')
    } catch { continue }

    // Find the function body for the entry point
    const fnBody = extractFunctionBody(source, cluster.entry)
    if (!fnBody) continue

    // Detect mutation patterns on function parameters
    const mutations = detectMutations(fnBody, cluster.entry)
    if (mutations.length > 0) {
      risks.push({ cluster: cluster.id, mutations })
    }
  }

  if (risks.length > 0) {
    const details = risks.map(r =>
      `${r.cluster}: ${r.mutations.join(', ')}`
    ).join('; ')
    return { ok: false, msg: `${risks.length} cluster(s) with argument mutation risk: ${details}` }
  }
  return { ok: true, msg: 'No argument mutation risks detected' }
}

function extractFunctionBody(source, functionName) {
  const patterns = [
    new RegExp(`(?:export\\s+)?(?:async\\s+)?function\\s+${escapeRegex(functionName)}\\s*\\([^)]*\\)\\s*\\{`, 'm'),
    new RegExp(`(?:export\\s+)?const\\s+${escapeRegex(functionName)}\\s*=\\s*(?:async\\s*)?\\([^)]*\\)\\s*=>\\s*\\{`, 'm'),
  ]

  for (const pattern of patterns) {
    const match = source.match(pattern)
    if (match) {
      let depth = 0
      let i = match.index + match[0].length - 1
      for (; i < source.length; i++) {
        if (source[i] === '{') depth++
        else if (source[i] === '}') {
          depth--
          if (depth === 0) break
        }
      }
      return source.slice(match.index + match[0].length, i)
    }
  }
  return null
}

function escapeRegex(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function detectMutations(fnBody, fnName) {
  const mutations = []

  // Pattern: delete arg.prop — mutating an argument by deleting a property
  // e.g., delete augmentedSchema.required (found in rjsf's getFirstMatchingOption)
  const deletePattern = /\bdelete\s+\w+\.\w+/g
  const deleteMatches = fnBody.match(deletePattern)
  if (deleteMatches) {
    mutations.push(`delete: ${[...new Set(deleteMatches)].join(', ')}`)
  }

  // Pattern: Object.assign(firstArg, ...) — mutates first argument
  const objectAssignPattern = /Object\.assign\(\s*\w+\s*,/g
  const assignMatches = fnBody.match(objectAssignPattern)
  if (assignMatches) {
    mutations.push(`Object.assign mutates first arg`)
  }

  // Pattern: arg.push(...) / arg.splice(...) / arg.sort() — mutating array arguments
  const arrayMutatePattern = /\w+\.(push|splice|sort|reverse|shift|pop|unshift)\s*\(/g
  const arrayMatches = fnBody.match(arrayMutatePattern)
  if (arrayMatches) {
    // Filter to only likely argument mutations (not local variable mutations)
    const unique = [...new Set(arrayMatches)]
    if (unique.length > 0) {
      mutations.push(`array mutation: ${unique.join(', ')}`)
    }
  }

  // Pattern: arg.prop = value — property assignment on what looks like a parameter
  // This is a heuristic — we check if the assigned object matches a function parameter name
  const propAssignPattern = /(\w+)\.\w+\s*=/g
  let propMatch
  const propAssignments = new Set()
  while ((propMatch = propAssignPattern.exec(fnBody)) !== null) {
    propAssignments.add(propMatch[1])
  }
  if (propAssignments.size > 0) {
    // Try to find parameter names from the function signature
    const paramMatch = fnBody.match && fnBody.match(/(?:function\s+\w+|const\s+\w+\s*=)\s*\(([^)]*)\)/)
    if (paramMatch) {
      const params = paramMatch[1].split(',').map(p => p.trim().split(':')[0].split('=')[0].trim()).filter(Boolean)
      const mutatingParams = [...propAssignments].filter(name => params.includes(name))
      if (mutatingParams.length > 0) {
        mutations.push(`param mutation: ${mutatingParams.map(p => `${p}.*=`).join(', ')}`)
      }
    }
  }

  return mutations
}

function checkChains() {
  const chainsJson = resolve(REGRET_DIR, 'chains.json')
  const chainsDir = resolve(REGRET_DIR, 'chains')

  if (!existsSync(chainsJson)) {
    return { ok: true, msg: 'No chains defined (optional)', warning: true }
  }

  try {
    const chains = JSON.parse(readFileSync(chainsJson, 'utf8'))
    const chainList = chains.chains || []
    if (chainList.length === 0) {
      return { ok: true, msg: 'No chains in chains.json', warning: true }
    }

    // Check that chain files exist
    if (!existsSync(chainsDir)) {
      return { ok: false, msg: `${chainList.length} chain(s) defined but no chains/ directory` }
    }

    let captured = 0
    let uncaptured = []
    for (const chain of chainList) {
      const chainFile = resolve(chainsDir, `${chain.id}.chain`)
      if (existsSync(chainFile)) {
        captured++
      } else {
        uncaptured.push(chain.id)
      }
    }

    if (uncaptured.length > 0) {
      return { ok: false, msg: `${uncaptured.length} chain(s) not captured: ${uncaptured.join(', ')}` }
    }
    return { ok: true, msg: `All ${captured} chain(s) captured` }
  } catch (e) {
    return { ok: false, msg: `Invalid chains.json: ${e.message}` }
  }
}

// ─── Main ──────────────────────────────────────────────────────────────────────

console.log('\n╔══════════════════════════════════════════════════════════════════╗')
console.log('║              REGRET PRE-REFACTOR AUDIT REPORT (JS/TS)           ║')
console.log('╚══════════════════════════════════════════════════════════════════╝')
console.log()

const checks = [
  ['Manifest', checkManifest],
  ['Regret Files', checkRegretFiles],
  ['Validation', checkValidate],
  ['Drift Detection', checkDrift],
  ['Cluster Health', checkHealth],
  ['Branch Coverage', checkCoverage],
  ['Mutation Risks', checkMutationRisks],
  ['Chains', checkChains],
]

let allPass = true
const warnings = []

for (const [name, checkFn] of checks) {
  process.stdout.write(`  Checking ${name}... `)
  try {
    const result = checkFn()
    const hasWarning = result.warning || false
    if (result.ok) {
      const icon = hasWarning ? '🟡' : '✅'
      console.log(`${icon} ${result.msg}`)
      if (hasWarning) warnings.push(`${name}: ${result.msg}`)
    } else {
      console.log(`❌ ${result.msg}`)
      allPass = false
    }
  } catch (e) {
    console.log(`⚠️  Error: ${e.message}`)
    warnings.push(`${name}: check failed — ${e.message}`)
  }
}

console.log()
console.log('─'.repeat(68))

if (allPass && warnings.length === 0) {
  console.log('✅ AUDIT PASSED — All checks GREEN. Safe to refactor.')
} else if (allPass && warnings.length > 0) {
  console.log('🟡 AUDIT PASSED with warnings:')
  for (const w of warnings) console.log(`   • ${w}`)
  console.log('   Refactoring is possible but consider addressing warnings first.')
} else {
  console.log('❌ AUDIT FAILED — Fix the issues above before refactoring.')
}

console.log()

if (strict && !allPass) {
  process.exit(1)
}
